### PR TITLE
paintbarn gold solution

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/usaco-923.mdx [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+author: Kevin Sheng
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_paintbarn_gold_feb19.html)
+
+## Explanation
+
+### Initial Observations
+
+**Note:** This solution assumes one has already had a firm grasp of the
```
Comment (envyaims) [inline solutions/usaco-923.mdx:14]: corny, maybe reword it ngl

File: solutions/usaco-923.mdx [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+[silver](http://www.usaco.org/current/data/sol_paintbarn_silver_feb19.html)
+version of this problem.
+
+We start off by calculating the layers of paint over each cell of the barn with
+prefix sums, but from there it's a bit hard to determine which layers we should
+paint. It's hard to directly calculate it from this array, but notice that there's
+precisely three states a cell can be in:
+
+1. If it's painted over, it will add $1$ to the area of optimal layers of paint.
```
Comment (envyaims) [inline solutions/usaco-923.mdx:23]: combine 1 and 2

File: solutions/usaco-923.mdx [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+	int optimal_amt;
+	read >> rect_num >> optimal_amt;
+
+	vector<vector<int>> barn(WIDTH, vector<int>(WIDTH));
+	for (int r = 0; r < rect_num; r++) {
+		int x1, y1, x2, y2;
+		read >> x1 >> y1 >> x2 >> y2;
+		for (int y = y1; y < y2; y++) {
+			barn[y][x1]++;
+			if (x2 < WIDTH) {
```
Comment (envyaims) [inline solutions/usaco-923.mdx:183]: wat this

**Review by envyaims — APPROVED** [review]
me gusta
just some minor things

File: solutions/usaco-923.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:14]: corny?

File: solutions/usaco-923.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:23]: they're different types?

File: solutions/usaco-923.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:183]: if the input is like
```
1 1 200 200
```
this program w/o this if statement will give an out of bounds error

File: solutions/usaco-923.mdx (hunk 1 above)
Comment (envyaims) [reply solutions/usaco-923.mdx:14]: reword it so you the average 3rd grader can understand it easier

File: solutions/usaco-923.mdx (hunk 2 above)
Comment (envyaims) [reply solutions/usaco-923.mdx:23]: I meant
"if its painted over, either it will add 1 or sub 1"

File: solutions/usaco-923.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/usaco-923.mdx:183]: why not just extend de array?

File: solutions/usaco-923.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:23]: will wait for third opinion

File: solutions/usaco-923.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:14]: cope

File: solutions/usaco-923.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:183]: because then it'd be a 201x201 array, and that's not what the barn is

Comment (envyaims) [reply solutions/usaco-923.mdx:183]: ok but like its weirder if you do this random operation that 99.5% of people avoid it by extending the array anyways

Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:183]: i can add a comment if you want

Comment (envyaims) [reply solutions/usaco-923.mdx:183]: @ryanchou-dev can you please convince sans its okay to extend the array by 1 and not stick so strict to the statement

Comment (envyaims) [reply solutions/usaco-923.mdx:183]: @SansPapyrus683 or youre better off using 0 indexing because youre like using 1 indexing on a 0 indexed array

File: solutions/usaco-923.mdx [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+to get the arrays with the following updated values:
+
+* $\texttt{top\_best[i]}$: the best matrix that lies ***at or above*** the $i$th row.
+* $\texttt{bottom\_best[i]}$: the best matrix that lies ***at or below*** $i$th row.
+* $\texttt{left\_best[i]}$: the best matrix that's ***at or to the left of*** the $i$th column.
+* $\texttt{right\_best[i]}$: the best matrix that's ***at or to the right of*** the $i$th column.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^3)$
```
Comment (dongliu0426) [inline solutions/usaco-923.mdx:153]: N is 10^5? do you mean max(x,y)^3?

File: solutions/usaco-923.mdx [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+	std::ifstream read("paintbarn.in");
+	int rect_num;
+	int optimal_amt;
+	read >> rect_num >> optimal_amt;
+
+	vector<vector<int>> barn(WIDTH, vector<int>(WIDTH));
+	for (int r = 0; r < rect_num; r++) {
+		int x1, y1, x2, y2;
+		read >> x1 >> y1 >> x2 >> y2;
+		for (int y = y1; y < y2; y++) {
```
Comment (dongliu0426) [inline solutions/usaco-923.mdx:181]: this is O(N*y), can you change it to O(N+xy) like the editorial (http://www.usaco.org/current/data/sol_paintbarn_gold_feb19.html)?

File: solutions/usaco-923.mdx (hunk 2 above)
Comment (dongliu0426) [reply solutions/usaco-923.mdx:23]: > I meant "if its painted over, either it will add 1 or sub 1"

yeah imo this makes more sense

File: solutions/usaco-923.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:153]: oh yeah

File: solutions/usaco-923.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:181]: i mean it still runs fine? don't think it has any big impact on readability either

Comment (dongliu0426) [reply solutions/usaco-923.mdx:181]: sure

File: solutions/usaco-923.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/usaco-923.mdx:183]: actually no, because
```
0 0 200 200
```
is valid input

Comment (ryanchou-dev) [reply solutions/usaco-923.mdx:185]: ok yall need to chill. it's js array indices. it only matters if it confuses the reader.

**Review by ryanchou-dev — APPROVED** [review]

**Review by dongliu0426 — APPROVED** [review]
